lib/hanami/application/routing/resolver.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Hanami
+  class Application
+    module Routing
+      # Hanami application router endpoint resolver
+      #
+      # @since 2.0.0
+      class Resolver
+        require "hanami/application/routing/resolver/trie"
+
+        # @api private
+        class NotCallableEndpointError < StandardError
+          def initialize(endpoint)
+            super("#{endpoint.inspect} isn't compatible with Rack. Please make sure it implements #call.")
+          end
+        end
+
+        attr_reader :container
+        attr_reader :base_namespace
+        attr_reader :inflector
+        attr_reader :slices
+
+        def initialize(container:, namespace:, inflector:, slices: Trie.new)
+          @container = container
+          @base_namespace = namespace
+          @inflector = inflector
+          @slices = slices
+        end
+
+        def with_container(new_container)
+          self.class.new(
+            container: new_container,
+            namespace: base_namespace,
+          )
+        end
+
+        # rubocop:disable Metrics/MethodLength
+        def call(path, identifier)
+          endpoint =
+            case identifier
+            when String
+              resolve_string_identifier(path, identifier)
+            when Class
+              identifier.respond_to?(:call) ? identifier : identifier.new
+            else
+              identifier
+            end
+
+          unless endpoint.respond_to?(:call) # rubocop:disable Style/IfUnlessModifier
+            raise NotCallableEndpointError.new(endpoint)
+          end
+
+          endpoint
+        end
+        # rubocop:enable Metrics/MethodLength
+
+        def register_slice(path, name)
+          slices.add(path, name)
+        end
+
+        private
+
+        def resolve_string_identifier(path, identifier)
+          slice_name = slices.find(path) or raise "missing slice for #{path.inspect} (#{identifier.inspect})"
+          slice = container.slices[slice_name]
+          action_key = "actions.#{identifier.gsub(/[#\/]/, '.')}"
```

Comment:
I want to make this centralized. Only one class should be responsible to hold those separators and provide private API to convert.

---

Making those separators configurable will increase the internal complexity (passing full-blown config), force other actors to be aware of it (e.g. CLI code generators), and grow the effort to communicate another setting.

I believe we should make a decision, and code this decision into a central place, by removing those inline values and operations (e.g. `#gsub`).

What do you think?

Reply:
Ohh, I think we might be misunderstanding each other here :)

What I was thinking should be configurable is the `"actions"` _prefix_ used when generating the identifier to get the action from the container.

For example, someone might want to make it `"web.actions"` (in fact, `"web.actions"` might be a better default) or even just `""` if they want to specify the absolute identifier in full within the route definitions.

So I wasn't talking about the _separators_ (i.e. the dots) but rather the _prefix_. I believe that could be pretty easily respected by generators and the like?

Reply:
@timriley OK, now I got what you mean.

I couldn't find in code where actions are registered. If possible, I want to make `actions.` prefix configurable and let the registration phase and the resolving phase to use this configuration.

There is no occurrence in code of `actions.` because they are auto-registered by `dry-container`? If so, we can't really make `actions.` configurable, right?

Reply:
@jodosha Actions will be auto-registered like most other components that go into the containers.

The reason to make the "actions base dir" a configurable thing is twofold:

1. When we have generators, the generators will respect that config and put the actions in that dir
2. And here in this resolver, it would use base dir to form the prefix of the container key, so the application authors don't need to repeat "actions.this" and "actions.that" and "actions.another_one" all the way through their routes file

Does that make sense?

I'm already doing similar with the views over in https://github.com/hanami/hanami/pull/1038, where I've added the `views.base_path` setting (which defaults to `"views"`). What we're talking about here is the equivalent config for actions.

Reply:
@timriley So the registration key depends on the directory. OK now it's clear. I'll adjust the code accordingly.

Reply:
@timriley I prefer to do this clean up when we'll have both the views and actions integrated.

Reply:
No worries, I can help with that. I’ll be revisiting the whole configuration anyway when I port it to dry-configurable